### ISSUE TITLE
interface (default) methods should show up

### ISSUE THREAD

**Issue by kares** [jruby/jruby#2735]
... on the Ruby side ... by alias-able etc.

esp. matters on Java 8 currently :

```
jruby-1.7.19 :004 > java.lang.Runnable.methods.grep /run/
 => [] 
```

**Comment by headius** [comment]
Known issue...and I had forgotten we needed to fix it. Will investigate.

**Comment by headius** [comment]
A bit of a snag here... we have attempted to add interface methods as empty bodies in the past, but they interfere with normal dispatch. For example, if you have a hierarchy like this:

```java
class Baz {
  public void blah() {}
}
interface Bar {
  public void blah();
}
class Foo implements Bar extends Baz { }
```

If the pseudo-module for "Bar" that we have in "Foo" has an empty body, the actual implementation in "Baz" will never be called (because it is masked by the default method). Default method dispatch is a bit more complicated because presumably javac/jvm will handle this case correctly with extra magic.

We may have to add logic into our hierarchy-building that adds the default methods only in hierarchies where they are not implemented by any class.

**Comment by headius** [comment]
FWIW, we have never populated methods on interfaces. I'm trying to think through it again and still having problems making it work. Interfaces just don't exist in the hierarchy of a Java class; they're only shapes that are applied on top of it, and so our module representation doesn't really match up exactly. Putting interface methods into those modules can't work because there are no methods in interfaces...only method signatures.

You say we need this...what do we need it for? Could we get away with having utility methods on our interface modules, a la Runnable.default_methods?

**Comment by kares** [comment]
right, reflection shows them as abstract thus not callable - no problem, but on Java 8 obtained default method handles (e.g. `java.util.Collection.class.getMethod("strean")`) will succeed when invoked with an instance passed as this ... so I was thinking (if possible) JRuby should reflect that scenario closely. 
I understand the limitation, even with "only" default methods showing up it's still an issue.

that is pretty much why I assumed we'll need those somehow implemented anyways, so that things like `java.util.Collection.java_class.to_java.getMethod("spliterator").invoke an_array_list` work easily when we do `method(:spliterator)` on the iface module ... with an `java.util.ArrayList` instance (overrides the default `spliterator` method) it still dispatches just like non-reflected Java would, so I was wrong that it would be useful.

if there's nothing you can think of we could do to improve here, feel free to close ...
